library/src/main/scala-new-collections/scala/scalajs/js/ArrayOps.scala

```diff
@@ -8,97 +8,1998 @@
+    xs
+  }
+
+  def dropInPlace(n: Int): js.Array[A] = {
+    xs.splice(0, normalized(n))
+    xs
+  }
+
+  @inline def dropRightInPlace(n: Int): js.Array[A] = {
+    trimEnd(n)
+    xs
+  }
+
+  def takeInPlace(n: Int): js.Array[A] = {
+    xs.length = normalized(n)
+    xs
+  }
+
+  def takeRightInPlace(n: Int): js.Array[A] = {
+    xs.splice(0, xs.length - normalized(n))
+    xs
+  }
 
-  // Scala notation for a fast concat()
+  def sliceInPlace(start: Int, end: Int): js.Array[A] =
+    takeInPlace(end).dropInPlace(start)
 
-  @inline def ++[B >: A](that: js.Array[_ <: B]): js.Array[B] =
-    ArrayOpsCommon.concat(array, that)
+  @inline private def normalized(n: Int): Int =
+    min(max(n, 0), xs.length)
 
-  // Methods whose inherited implementations do not play nice with the optimizer
+  def dropWhileInPlace(p: A => Boolean): js.Array[A] = {
+    val idx = indexWhere(!p(_))
+    if (idx < 0) {
+      clear()
+      xs
+    } else {
+      dropInPlace(idx)
+    }
+  }
 
-  @inline override def reduceLeft[B >: A](op: (B, A) => B): B =
-    ArrayOpsCommon.reduceLeft(array, op)
+  def takeWhileInPlace(p: A => Boolean): js.Array[A] = {
+    val idx = indexWhere(!p(_))
+    if (idx < 0)
+      xs
+    else
+      takeInPlace(idx)
+  }
 
-  @inline override def reduceRight[B >: A](op: (A, B) => B): B =
-    ArrayOpsCommon.reduceRight(array, op)
+  def padToInPlace(len: Int, elem: A): js.Array[A] = {
+    while (xs.length < len)
+      xs.push(elem)
+    xs
+  }
 
-  // Inspired by mutable.IndexedOptimizedSeq
+  // IndexedBuffer[A]
 
-  def mapInPlace(f: A => A): this.type = {
+  def flatMapInPlace(f: A => IterableOnce[A]): js.Array[A] = {
+    // There's scope for a better implementation which copies elements in place.
     var i = 0
-    val size = this.size
-    while (i < size) {
-      this(i) = f(this(i))
+    val copy = xs.jsSlice()
+    xs.length = 0
+    for {
+      x <- copy
+      y <- f(x).iterator
+    } {
+      xs.push(y)
+    }
+    xs
+  }
+
+  def filterInPlace(p: A => Boolean): js.Array[A] = {
+    var i = 0
+    var j = 0
+    val len = xs.length
+    while (i < len) {
+      if (p(xs(i))) {
+        if (i != j) {
+          xs(j) = xs(i)
+        }
+        j += 1
+      }
       i += 1
     }
-    this
+
+    if (i != j)
+      xs.length = j
+    xs
   }
 
 }
 
-private object ArrayOps {
-  private object ArrayFactory extends SeqFactory[js.Array] {
-    @inline
-    def empty[A]: js.Array[A] = js.Array()
+object ArrayOps {
+  @SerialVersionUID(3L)
+  private class ArrayView[A](xs: js.Array[A]) extends IndexedSeqView[A] {
+    @inline def length: Int = xs.length
 
-    @inline
-    def from[A](source: IterableOnce[A]): js.Array[A] =
-      (newBuilder[A] ++= source).result()
+    @inline def apply(n: Int): A = xs(n)
 
-    @inline
-    def newBuilder[A]: mutable.Builder[A, js.Array[A]] =
-      new ArrayOps[A]
+    override protected[this] def className = "ArrayView"
```

Comment:
Shouldn't this return a JS specific prefix? Otherwise `toString` will not be distinguishable from a `scala.Array` view (I suppose).